math/ncr_modulo_p.cpp

```diff
@@ -0,0 +1,100 @@
+/**
+ * @file
+ * @brief This program aims at calculating nCr modulo p
+ *
+ */
+
+#include <cassert>
+#include <iostream>
+#include <vector>
+
+/** Finds the value of x, y such that a*x + b*y = gcd(a,b)
+ *
+ * @params[in] the numbers 'a', 'b' and address of 'x' and 'y' from above
+ * equation
+ * @returns the gcd of a and b
+ */
+int64_t gcdExtended(int64_t a, int64_t b, int64_t *x, int64_t *y) {
+    if (a == 0) {
+        *x = 0, *y = 1;
+        return b;
+    }
+
+    int64_t x1 = 0, y1 = 0;
+    int64_t gcd = gcdExtended(b % a, a, &x1, &y1);
+
+    *x = y1 - (b / a) * x1;
+    *y = x1;
+    return gcd;
+}
+
+/** Find modular inverse of a with m i.e. a number x such that (a*x)%m = 1
+ *
+ * @params[in] the numbers 'a' and 'm' from above equation
+ * @returns the modular inverse of a
+ */
+int64_t modInverse(int64_t a, int64_t m) {
+    int64_t x = 0, y = 0;
+    int64_t g = gcdExtended(a, m, &x, &y);
+    if (g != 1) {  // modular inverse doesn't exist
+        return -1;
+    }
+    else {
+        int64_t res = (x % m + m) % m;
+        return res;
+    }
+}
+
+std::vector<int64_t> fac;
```

Comment:
Avoid using global variables, they may cause overflows and other issues.
Pass them as local function parameters, or as class private members.

Reply:
what if I keep that variable inside namespace?
And later access them as math::fac?

Reply:
Easiest thing you can do is to pass them as local function parameters.
Or even better, add them as (public) class members, and call the variable directly to the function.

Reply:
Yes, I refactored the code so that the factorial vector and the remaining functions are inside a class